Rules/ReviewUnusedParameter.cs

```diff
@@ -0,0 +1,129 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Management.Automation.Language;
+using Microsoft.Windows.PowerShell.ScriptAnalyzer.Generic;
+#if !CORECLR
+using System.ComponentModel.Composition;
+#endif
+using System.Globalization;
+
+namespace Microsoft.Windows.PowerShell.ScriptAnalyzer.BuiltinRules
+{
+    /// <summary>
+    /// ReviewUnusedParameter: Check that all declared parameters are used in the script body.
+    /// </summary>
+#if !CORECLR
+[Export(typeof(IScriptRule))]
+#endif
+    public class ReviewUnusedParameter : IScriptRule
+    {
+        public IEnumerable<DiagnosticRecord> AnalyzeScript(Ast ast, string fileName)
+        {
+            if (ast == null)
+            {
+                throw new ArgumentNullException(Strings.NullAstErrorMessage);
+            }
+
+            var scriptBlockAsts = ast.FindAll(oneAst => oneAst is ScriptBlockAst, true);
+            if (scriptBlockAsts == null)
+            {
+                yield break;
+            }
+
+            foreach (ScriptBlockAst scriptBlockAst in scriptBlockAsts)
+            {
+                // find all declared parameters
+                IEnumerable<Ast> parameterAsts = scriptBlockAst.FindAll(oneAst => oneAst is ParameterAst, false);
+
+                // list all variables
+                IEnumerable<string> variables = scriptBlockAst.FindAll(oneAst => oneAst is VariableExpressionAst, false)
+                    .Cast<VariableExpressionAst>()
+                    .Select(variableExpressionAst => variableExpressionAst.VariablePath.ToString());
+
+                foreach (ParameterAst parameterAst in parameterAsts)
+                {
+                    // compare the list of variables to the parameter name
+                    // there should be at least two matches of the variable name since the parameter declaration counts as one
+                    int matchCount = variables
+                        .Where(variable => variable == parameterAst.Name.VariablePath.ToString())
+                        .Take(2).Count();
+                    if (matchCount == 2)
+                    {
+                        continue;
+                    }
+
+                    // all bets are off if the script uses PSBoundParameters
+                    if (variables.Contains("PSBoundParameters"))

```

Comment:
No need to do this in the foreach. Instead we should do it as soon as we have the set of variables.

Reply:
fixed in 938a2835ac8d3c8e5470642d923c3b85476b303c